### ISSUE TITLE
training time

### ISSUE THREAD

**Issue by honey218518** [Glasssix/joint_bayesian#1]
How long is the training process? When I use the lbp_lfw, calculating Su Sw for once takes more than one day. The code is followed: 
//初始化Su，Sw   Su=cov(u), Sw=cov(e)
    for (int i = 0; i < u.rows(); i++)
        for (int j = 0; j < u.rows(); j++){
            Matrix<double, Dynamic, Dynamic>xi, xj;
            xi = u.row(i);
            xj = u.row(j);
            Su(i, j) = cov(xi, xj);
            xi = e.row(i);
            xj = e.row(j);
            Sw(i, j) = cov(xi, xj);
        }

**Comment by xiaoma340** [comment]
1.You can't use the original image directly.
2.Image must be reduced dimension.
3.My image feature array is 1*40,I use 391908 images to train joint-bayesian,feature array is 391908*40,that cost 15 minutes and it will  converge in 4 iters.
4.Can you speak Chinese，I'm poor  at English.

**Comment by honey218518** [comment]
你的意思是将lbp特征先从5900维降到40维？你实验的时候准确率有多少？另外你的代码在迭代的过程中u的值没有改变，ui += SuFG(everyclasscount[i])_(dataset.row(count1 + j).transpose()); 循环完了之后是不是要将其给u，即
for (int j = 0; j < everyclasscount[i]; j++)ui += SuFG(everyclasscount[i])_(dataset.row(count1 + j).transpose());改为
for (int j = 0; j < everyclasscount[i]; j++)
{
    ui += SuFG(everyclasscount[i])*(dataset.row(count1 + j).transpose());
}
u.col(i) = ui;

**Comment by xiaoma340** [comment]
```
  是的是的，要改过来，我菜鸟一个，前面训练的时候也能收敛，根本没发现这个错误，非常感谢。
  改过来之后，重新测试，正确率是88.6%。你真心厉害，以后我也要仔细检查代码了。
  维度一定要降，不然量太大了，没法算，我觉得不用降到40维这么低，可能丢失很多信息。
```

**Comment by inlmouse** [comment]
@honey218518 I prepared the training data, so let me answer your question. We used a modified version of DeepID3, and extract single image(64*64*3)'s feature into 40-dimensions. In [Tang's paper](https://arxiv.org/pdf/1502.00873v1.pdf), this 40-dimensions feature will obtain a high robustness feature expression ability, both in numerical value and binary pattern. The using of LBP feature is not recommended.

**Comment by honey218518** [comment]
@xiaomage340 你的代码是不是改了？

**Comment by honey218518** [comment]
@inlmouse 请问你实现过deepid1没？

**Comment by inlmouse** [comment]
@honey218518 Yes, we‘ve implemented DeepID2+ without data patch.

**Comment by honey218518** [comment]
@inlmouse 我现在在实验deepid1，麻烦你帮我看下我这个有没有啥问题。用的casia数据集，部分图片（能检测出人脸的且人数>=5），未对齐，只用了一个patch；人数为10497，训练图片321124，验证图片29311，图片大小39*39的灰度图。模型对验证集的准确是57%。

**Comment by inlmouse** [comment]
If you can detect more than 1 face in 1 image, change your method and re-detect. Make sure that only 1 face are cropped for further processing. Secondly, alignment is recommended. It will promote about 10% on LFW(our implementation is not official in the strict sense, so 10% is also an empirical value). And in my opinion, 39*39 is a little small considering that you did not patch. For more detail, your network.prototxt is required.

**Comment by honey218518** [comment]
@inlmouse 网络结构如下：
#############################  DATA Layer  #############################
name: "face_train_val"
layer {
  top: "data_1"
  top: "label_1"
  name: "data_1"
  type: "Data"
  data_param {
    source: "examples/deepid/deepid_train_lmdb"
    backend:LMDB
    batch_size: 128
  }
  transform_param {
     mean_file: "examples/deepid/deepid_lmdb_mean.binaryproto"
     mirror: true
  }
  include: { phase: TRAIN }
}

layer {
  top: "data_1"
  top: "label_1"
  name: "data_1"
  type: "Data"
  data_param {
    source: "examples/deepid/deepid_test_lmdb"
    backend:LMDB
    batch_size: 128
  }
  transform_param {
    mean_file: "examples/deepid/deepid_lmdb_mean.binaryproto"
    mirror: true
  }
  include: { 
    phase: TEST 
  }
}

#############################  CONV NET 1 #############################
layer {
  name: "conv1_1"
  type: "Convolution"
  bottom: "data_1"
  top: "conv1_1"
  param {
    name: "conv1_w"
    lr_mult: 1
    decay_mult: 1
  }
  param {
    name: "conv1_b"
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 20
    kernel_size: 4
    stride: 1
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "relu1_1"
  type: "ReLU"
  bottom: "conv1_1"
  top: "conv1_1"
}
layer {
  name: "norm1_1"
  type: "LRN"
  bottom: "conv1_1"
  top: "norm1_1"
  lrn_param {
    local_size: 5
    alpha: 0.0001
    beta: 0.75
  }
}
layer {
  name: "pool1_1"
  type: "Pooling"
  bottom: "norm1_1"
  top: "pool1_1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv2_1"
  type: "Convolution"
  bottom: "pool1_1"
  top: "conv2_1"
  param {
    name: "conv2_w"
    lr_mult: 1
    decay_mult: 1
  }
  param {
    name: "conv2_b"
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 40
    kernel_size: 3
    #group: 2
    stride: 1
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }

}
layer {
  name: "relu2_1"
  type: "ReLU"
  bottom: "conv2_1"
  top: "conv2_1"
}
layer {
  name: "norm2_1"
  type: "LRN"
  bottom: "conv2_1"
  top: "norm2_1"
  lrn_param {
    local_size: 5
    alpha: 0.0001
    beta: 0.75
  }
}
layer {
  name: "pool2_1"
  type:  "Pooling"
  bottom: "norm2_1"
  top: "pool2_1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv3_1"
  type: "Convolution"
  bottom: "pool2_1"
  top: "conv3_1"
  param {
    name: "conv3_w"
    lr_mult: 1
    decay_mult: 1
  }
  param {
    name: "conv3_b"
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 60
    kernel_size: 3
    stride: 1
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }

}
layer {
  name: "pool3_1"
  type:  "Pooling"
  bottom: "conv3_1"
  top: "pool3_1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "conv4_1"
  type: "Convolution"
  bottom: "pool3_1"
  top: "conv4_1"
  param {
    name: "conv4_w"
    lr_mult: 1
    decay_mult: 1
  }
  param {
    name: "conv4_b"
    lr_mult: 2
    decay_mult: 0
  }
  convolution_param {
    num_output: 80
    kernel_size: 2
    stride: 2
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }

}
layer{
  name:"flatten_pool3_1"
  type:"Flatten"
  bottom:"pool3_1"
  top:"flatten_pool3_1"
}
layer{
  name:"flatten_conv4_1"
  type:"Flatten"
  bottom:"conv4_1"
  top:"flatten_conv4_1"
}
layer{
  name:"contact_conv"
  type:"Concat"
  bottom:"flatten_conv4_1"
  bottom:"flatten_pool3_1"
  top:"contact_conv"
}
layer {
  name: "deepid_1"
  type:  "InnerProduct"
  bottom: "contact_conv"
  top: "deepid_1"
  param {
    name: "fc6_w"
    lr_mult: 1
    decay_mult: 1
  }
  param {
    name: "fc6_b"
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 160
    weight_filler {
      type: "gaussian"
      std: 0.005
    }
    bias_filler {
      type: "constant"
      value: 0.1
    }
  }

}
layer {
  name: "relu6_1"
  type: "ReLU"
  bottom: "deepid_1"
  top: "deepid_1"
}
layer {
  name: "drop6_1"
  type:  "Dropout"
  bottom: "deepid_1"
  top: "deepid_1"
  dropout_param {
    dropout_ratio: 0.5
  }
}

layer {
  name: "fc8_1"
  type:  "InnerProduct"
  bottom: "deepid_1"
  top: "fc8_1"
  param {
    name: "fc8_w"
    lr_mult: 1
    decay_mult: 1
  }
  param {
    name: "fc8_b"
    lr_mult: 2
    decay_mult: 0
  }
  inner_product_param {
    num_output: 10497
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layer {
  name: "accuracy_1"
  type:  "Accuracy"
  bottom: "fc8_1"
  bottom: "label_1"
  top: "accuracy_1"
  include: { phase: TEST }
}
layer {
  name: "loss_1"
  type:  "SoftmaxWithLoss"
  bottom: "fc8_1"
  bottom: "label_1"
  top: "loss_1"
  #loss_weight: 0.5
}
我打算的先测试一个版本，然后再加对齐，图片归一化时尺寸变大，增加patch等。deep1原文种验证集的正确率也是50%+，但是我在网上看到有的人训练验证集的准确率 >70%，我就不太确定了。。另外，在实现联合贝叶斯的时候，casia数据集我都用来进行deepid训练，可不可以就用casia的数据集训练联合贝叶斯？谢谢哈

**Comment by inlmouse** [comment]
@honey218518 I'll except your testing result, keep in touch : ). I have the same question. Some proclaimed that they are able to get over 90% on CASIA classification task. However, we only obtained 84% on a nonalignment CASIA. Choosing 1500 samples of CASIA to Joint-Bayesian is most effective. Using the whole dataset only get 1~2% performance promotion on LFW.

**Comment by honey218518** [comment]
@inlmouse 你的84%是指的deepid训练验证准确率还是后续联合贝叶斯的准确率

**Comment by inlmouse** [comment]
@honey218518 I mean CASIA classification task we can obtain 84%, not LFW.

**Comment by honey218518** [comment]
@inlmouse 你说的classification task 84%是指的文章里面10%的validation data的准确率，还是所有CASIA数据集的分类准确率

**Comment by inlmouse** [comment]
@honey218518  validation data accuracy, but it also contained 10575 samples. Our implementation is not official, something maybe different from the paper.

**Comment by honey218518** [comment]
@inlmouse 我现在在训练联合贝叶斯，1500人，85756张图片，特征维度160，估计会训练多久？初始化Su、Sw目测需要两三个小时。。。

**Comment by xiaoma340** [comment]
刚刚上传了更新后的工程，接口不变，训练速度大大提升，你用更新后的吧。

**Comment by honey218518** [comment]
@xiaomage340 嗯，你把计算协方差的部分改了，我之前看到这样的一段话“原来协方差矩阵还可以这样计算，先让样本矩阵中心化，即每一维度减去该维度的均值，使每一维度上的均值为0，然后直接用新的到的样本矩阵乘上它的转置，然后除以(N-1)即可。”，应该是你新的版本的意思吧

**Comment by xiaoma340** [comment]
是的，eigen库矩阵运算很快，这方法比之前的协方差计算方法快了几十倍，调试的时候中间加了十几个cout忘了删了，你自己删了吧

**Comment by honey218518** [comment]
@inlmouse 我刚测试了几十张明显图片，准确率只有57%。。。帮我用你做的测试下，图片我传到百度云
链接：http://pan.baidu.com/s/1jHIjx7g 密码：5jot

**Comment by honey218518** [comment]
@xiaomage340 麻烦你也帮我测试下链接：http://pan.baidu.com/s/1jHIjx7g 密码：5jot

**Comment by honey218518** [comment]
@xiaomage340 @inlmouse 我用casia数据集，选了8902个人，每个人2张，总共组成了8902对同一个人的和8902对不是同一个人的，准确率是82.98%

**Comment by xiaoma340** [comment]
你这正确率已经很高了啊

**Comment by honey218518** [comment]
@xiaomage340 但是我测的我自己的数据集就只有57%

**Comment by honey218518** [comment]
@inlmouse 请问下人脸对齐你是怎么做的？文章中做的相似变换，也就是旋转就行了？

**Comment by xiaoma340** [comment]
先检测关键点，然后根据检测出的双眼关键点，用opencv进行仿射变换对齐
